subprojects/GOPATH/src/github.com/microsoft/hivedscheduler/pkg/algorithm/hived_algorithm.go

```diff
@@ -44,10 +44,20 @@ type HivedAlgorithm struct {
 	vcSchedulers map[api.VirtualClusterName]intraVCScheduler
 	// scheduler for opportunistic pods
 	opportunisticSchedulers map[CellChain]*topologyAwareScheduler
-	// ChainCellLists of physical cells of each cell chain
+	// ChainCellLists of physical cells of each cell chain (including the children of the free cells)
 	fullCellList map[CellChain]ChainCellList
 	// ChainCellLists of free physical cells of each cell chain (used in buddy alloc)
 	freeCellList map[CellChain]ChainCellList
+	// remaining quota of each VC of each cell type
+	vcRemainingQuotas map[api.VirtualClusterName]map[CellChain]map[CellLevel]int32
+	// total remaining quota of each cell type
+	totalRemainingQuotas map[CellChain]map[CellLevel]int32
+	// number of cells left in the physical cluster of each cell type
+	// (i.e., the cell is either in the free list or could be obtained by splitting
+	// such a free cell. Note that these cells include both healthy and bad cells)
+	numLeftCells map[CellChain]map[CellLevel]int32
+	// number of bad cells in the physical cluster of each cell type
+	numBadCells map[CellChain]map[CellLevel]int32
```

Comment:
Could a bad cell is also being used by a job?

Reply:
Seems it could be. Let's comment at CellBadState

---
In reply to: [388131306](https://github.com/microsoft/pai/pull/4153#discussion_r388131306) [](ancestors = 388131306)

Reply:
Solved as we have used separated states for bad and used